### ISSUE TITLE
tasks/ssl.yml craps out at line 16

### ISSUE THREAD

**Issue by relumer** [cmacrae/ansible-sensu#56]
Hello
i have this problem , where the play stops because it can not copy ssl cers. Some help?

tasks/ssl_generate.yml:7
skipping: [debian3] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
skipping: [debian5] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}

TASK [cmacrae.sensu : Untar the ssl_certs tarball from sensuapp.org] ***********
task path: /home/dimtheo/my-roles/cmacrae.sensu/tasks/ssl_generate.yml:17
skipping: [debian3] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
skipping: [debian5] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}

TASK [cmacrae.sensu : Generate SSL certs] **************************************
task path: /home/dimtheo/my-roles/cmacrae.sensu/tasks/ssl_generate.yml:25
skipping: [debian3] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
skipping: [debian5] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}

TASK [cmacrae.sensu : Stash the Sensu SSL certs/keys] **************************
task path: /home/dimtheo/my-roles/cmacrae.sensu/tasks/ssl_generate.yml:36
skipping: [debian3] => (item=sensu_ca/cacert.pem)  => {"changed": false, "item": "sensu_ca/cacert.pem", "skip_reason": "Conditional check failed", "skipped": true}
skipping: [debian5] => (item=sensu_ca/cacert.pem)  => {"changed": false, "item": "sensu_ca/cacert.pem", "skip_reason": "Conditional check failed", "skipped": true}
skipping: [debian5] => (item=server/cert.pem)  => {"changed": false, "item": "server/cert.pem", "skip_reason": "Conditional check failed", "skipped": true}
skipping: [debian3] => (item=server/cert.pem)  => {"changed": false, "item": "server/cert.pem", "skip_reason": "Conditional check failed", "skipped": true}
skipping: [debian5] => (item=server/key.pem)  => {"changed": false, "item": "server/key.pem", "skip_reason": "Conditional check failed", "skipped": true}
skipping: [debian3] => (item=server/key.pem)  => {"changed": false, "item": "server/key.pem", "skip_reason": "Conditional check failed", "skipped": true}
skipping: [debian5] => (item=client/cert.pem)  => {"changed": false, "item": "client/cert.pem", "skip_reason": "Conditional check failed", "skipped": true}
skipping: [debian3] => (item=client/cert.pem)  => {"changed": false, "item": "client/cert.pem", "skip_reason": "Conditional check failed", "skipped": true}
skipping: [debian5] => (item=client/key.pem)  => {"changed": false, "item": "client/key.pem", "skip_reason": "Conditional check failed", "skipped": true}
skipping: [debian3] => (item=client/key.pem)  => {"changed": false, "item": "client/key.pem", "skip_reason": "Conditional check failed", "skipped": true}

 Deploy the Sensu client SSL cert/key] ********************
task path: /home/dimtheo/my-roles/cmacrae.sensu/tasks/ssl.yml:16
[DEPRECATION WARNING]: Skipping task due to undefined Error, in the future this will be a fatal error.: {{ sensu_ssl_tool_base_path }}/client/cert.pem: {{ 
dynamic_data_store }}/{{ groups['sensu_masters'][0] }}{{ sensu_master_config_path }}/ssl_generation/sensu_ssl_tool: {{ 
hostvars[groups['sensu_masters'][0]]['sensu_config_path'] | default('/etc/sensu') }}: 'dict object' has no attribute 'sensu_masters'.
This feature will be 
removed in a future release. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
[DEPRECATION WARNING]: Skipping task due to undefined Error, in the future this will be a fatal error.: {{ sensu_ssl_tool_base_path }}/client/cert.pem: {{ 
dynamic_data_store }}/{{ groups['sensu_masters'][0] }}{{ sensu_master_config_path }}/ssl_generation/sensu_ssl_tool: {{ 
hostvars[groups['sensu_masters'][0]]['sensu_config_path'] | default('/etc/sensu') }}: 'dict object' has no attribute 'sensu_masters'.
This feature will be 
removed in a future release. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
fatal: [debian3]: FAILED! => {"failed": true, "msg": "'item' is undefined"}
fatal: [debian5]: FAILED! => {"failed": true, "msg": "'item' is undefined"}

**Comment by relumer** [comment]
if i run the code with the parameters like this
{ role: cmacrae.sensu, sensu_master: true, sensu_include_dashboard: true  }
it stops at this point :+1: 
tasks/ssl_generate.yml:36
fatal: [debian3]: FAILED! => {"failed": true, "msg": "{{ ansible_conf_path }}/data/dynamic: 'ansible_conf_path' is undefined"}
fatal: [debian5]: FAILED! => {"failed": true, "msg": "{{ ansible_conf_path }}/data/dynamic: 'ansible_conf_path' is undefined"

my understanding is that this has to do with these lines 

 src: "{{ sensu_config_path }}/ssl_generation/sensu_ssl_tool/{{ item }}"
      dest: "{{ dynamic_data_store }}"


what am i missing ? what files do i need to have in that directory? dynamic_data_store: "{{ ansible_conf_path }}/data/dynamic"           ?    
can someone help ? thanks   @cmacrae @smuth4

**Comment by relumer** [comment]
ok i managed to get past this datastore error , i figured it out but it still stops here


TASK [cmacrae.sensu : Deploy the Sensu client SSL cert/key] ********************
task path: /home/dimtheo/my-roles/cmacrae.sensu/tasks/ssl.yml:16
fatal: [debian3]: FAILED! => {"failed": true, "msg": "'item' is undefined"}
fatal: [debian1]: FAILED! => {"failed": true, "msg": "'item' is undefined"}
fatal: [debian5]: FAILED! => {"failed": true, "msg": "'item' is undefined"}

**Comment by smuth4** [comment]
I can only reproduce this by installing ansible version 2.0.0 and modifying ssl.yml to remove the with_items for that task. Can you confirm you're on the latest version of this repo and ssl.yml is unmodified?

**Comment by relumer** [comment]
Hi
This is my ansible version 
ansible --version
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

and this is my ssl.yml , unmodified

 - name: Deploy the Sensu client SSL cert/key
    copy:
      src: "{{ item.src }}"
      owner: "{{ sensu_user_name }}"
      remote_src: "{{ sensu_ssl_deploy_remote_src }}"
      group: "{{ sensu_group_name }}"
      dest: "{{ sensu_config_path }}/ssl/{{ item.dest }}"
    with_items:
      - {src: "{{ sensu_ssl_client_cert }}", dest: cert.pem}
      - {src: "{{ sensu_ssl_client_key }}", dest: key.pem}
    notify: restart sensu-client service


i got the role from galaxy , is this the latest?

**Comment by relumer** [comment]
i decided to create few new debian 8 VMs to try this role again
it stops at the same file again

tasks/ssl.yml:16
fatal: [debian10]: FAILED! => {"failed": true, "msg": "the file_name '/etc/ansible/data/dynamic/debian7/etc/sensu/ssl_generation/sensu_ssl_tool/client/cert.pem' does not exist, or is not readable"}
fatal: [debian9]: FAILED! => {"failed": true, "msg": "the file_name '/etc/ansible/data/dynamic/debian7/etc/sensu/ssl_generation/sensu_ssl_tool/client/cert.pem' does not exist, or is not readable"}
fatal: [debian11]: FAILED! => {"failed": true, "msg": "the file_name '/etc/ansible/data/dynamic/debian7/etc/sensu/ssl_generation/sensu_ssl_tool/client/cert.pem' does not exist, or is not readable"}
fatal: [debian12]: FAILED! => {"failed": true, "msg": "the file_name '/etc/ansible/data/dynamic/debian7/etc/sensu/ssl_generation/sensu_ssl_tool/client/cert.pem' does not exist, or is not readable"}
fatal: [debian8]: FAILED! => {"failed": true, "msg": "the file_name '/etc/ansible/data/dynamic/debian7/etc/sensu/ssl_generation/sensu_ssl_tool/client/cert.pem' does not exist, or is not readable"}
fatal: [debian7]: FAILED! => {"failed": true, "msg": "the file_name '/etc/ansible/data/dynamic/debian7/etc/sensu/ssl_generation/sensu_ssl_tool/client/cert.pem' does not exist, or is not readable"}

any ideas.. ? should i just comment out the with_items  and the 2 src: lines ?

**Comment by relumer** [comment]
ok i commented them out , now it gives these errors

/tasks/ssl.yml:16
fatal: [debian8]: FAILED! => {"failed": true, "msg": "'item' is undefined"}
fatal: [debian10]: FAILED! => {"failed": true, "msg": "'item' is undefined"}
fatal: [debian9]: FAILED! => {"failed": true, "msg": "'item' is undefined"}
fatal: [debian12]: FAILED! => {"failed": true, "msg": "'item' is undefined"}
fatal: [debian11]: FAILED! => {"failed": true, "msg": "'item' is undefined"}
fatal: [debian7]: FAILED! => {"failed": true, "msg": "'item' is undefine




`

**Comment by relumer** [comment]
i think i should not comment out those lines because i have set the machine debian7 to be a sensu master . So , what is causing this problem ?

it does not create the necessary folder and files named debian7 in etc/ansible/data/dynamic/

**Comment by smuth4** [comment]
Does debian7 have sensu_master set to true?

**Comment by relumer** [comment]
Hi
let me rephrase what i just said
I am pushing the code from Ubuntu to 6 new debian 8 vms. I have set one of those machines , named debian7 to act as the master in my hosts file like this 
[sensu_masters]
debian7 


am i missing something?

**Comment by smuth4** [comment]
yes, a host also needs to have sensu_master set to true to act as the master. If this is a closed system, you'll also want to set rabbitmq_server and redis_server to true for debian7 as well.

**Comment by relumer** [comment]
so i should install ansible on debian7 and rerun the code from there again , with the approrpiate changes in defaults/main,yml   

ok i will be back soon with news

**Comment by relumer** [comment]
ok it seems to be working , it passed that task file. I set debian7 to be sensu master and rabbitmq true and it started deploying code with no errors

it is now stuck for 45min at this point

"Setting up socat (1.7.2.4-2) ...", 
        "Setting up rabbitmq-server (3.6.6-1) ...", 
        "Adding group `rabbitmq' (GID 114) ...", 
        "Done.", 
        "Adding system user `rabbitmq' (UID 110) ...", 
        "Adding new user `rabbitmq' (UID 110) with group `rabbitmq' ...", 
        "Not creating home directory `/var/lib/rabbitmq'.", 
        "Processing triggers for libc-bin (2.19-18+deb8u6) ...", 
        "Processing triggers for systemd (215-17+deb8u5) ..."
    ]
}

**Comment by relumer** [comment]
its been like that for hours now , it doesnt finish but doesn't stop either  "Processing triggers for libc-bin (2.19-18+deb8u6) ...", 
        "Processing triggers for systemd (215-17+deb8u5) ..."
    ]
}
 and a blinking cursor

**Comment by cmacrae** [comment]
@relumer It's times like these you need to start digging to find out what the issue might be. This seems like `apt` is waiting/stuck on processing package instructions. So, you should probably SSH into the system and inspect the process using `ps`. Probably a good idea to check logs and general resources of that system, too - and, if necessary, attach to the process using something like `strace` if nothing obvious comes up from these other general troubleshooting steps.  

Whilst I appreciate you're likely eager to get this up and running, this latest problem doesn't look at all related to this Ansible role, so we can't really be expected to help you with basic troubleshooting.

One last thing I will say: make sure you've got the latest package lists cached on your system (`apt-get update`) before starting this.

**Comment by relumer** [comment]
hi @cmacrae  the only running processes regarding the role are these 

ps -ef | grep apt-get
root     25531 25525  0 12:28 pts/4    00:00:00 grep apt-get
root@debian7:/home/dimtheo# ps -ef | grep apt
root      6030  5993  0 Dec08 pts/0    00:00:00 sshpass -d24 ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r -tt 192.168.1.96 /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1481228034.84-24233648838519/apt.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1481228034.84-24233648838519/" > /dev/null 2>&1 && sleep 0'
root      6037  6030  0 Dec08 pts/3    00:00:00 ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r -tt 192.168.1.96 /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1481228034.84-24233648838519/apt.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1481228034.84-24233648838519/" > /dev/null 2>&1 && sleep 0'
root      6040  5994  0 Dec08 pts/0    00:00:00 sshpass -d24 ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r -tt 192.168.1.98 /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1481228035.68-75551965922814/apt.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1481228035.68-75551965922814/" > /dev/null 2>&1 && sleep 0'
root      6044  6040  0 Dec08 pts/6    00:00:00 ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r -tt 192.168.1.98 /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1481228035.68-75551965922814/apt.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1481228035.68-75551965922814/" > /dev/null 2>&1 && sleep 0'
root      6049  5995  0 Dec08 pts/0    00:00:00 sshpass -d24 ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r -tt 192.168.1.95 /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1481228036.89-182386246727374/apt.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1481228036.89-182386246727374/" > /dev/null 2>&1 && sleep 0'
root      6050  6049  0 Dec08 pts/7    00:00:00 ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r -tt 192.168.1.95 /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1481228036.89-182386246727374/apt.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1481228036.89-182386246727374/" > /dev/null 2>&1 && sleep 0'
root      6052  5996  0 Dec08 pts/0    00:00:00 sshpass -d24 ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r -tt 192.168.1.99 /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1481228037.34-116466212401969/apt.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1481228037.34-116466212401969/" > /dev/null 2>&1 && sleep 0'
root      6054  6052  0 Dec08 pts/10   00:00:00 ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o User=root -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r -tt 192.168.1.99 /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1481228037.34-116466212401969/apt.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1481228037.34-116466212401969/" > /dev/null 2>&1 &&



i understand you should not debug my systems but do these have something to do only with the system and not the role?

**Comment by cmacrae** [comment]
@relumer The tasks running are simple executions of `apt` commands. I'd say a good way of debugging would be to get the system to the exact same state it was before running this Ansible role against it - if you're using VMs, deploy anew, then snapshot before applying the Ansible role - and see if you're able to carry out the same `apt` actions manually, without any intervention outside of what Ansible is doing. That should help determine if it's simply a misconfiguration with your system.

**Comment by relumer** [comment]
ok @cmacrae i will redo everything in a new vm. 
i only had to install sshpass for ansible ssh to work  and i see this maybe causing the issues
but i will redo everything

**Comment by cmacrae** [comment]
Cheers @relumer, good luck 👍🏻
